lsp/nls/src/linearization/mod.rs

```diff
@@ -247,35 +260,72 @@ impl Linearizer for AnalysisHost {
+                        Term::Op2(_, rt1, rt2) => {
+                            walk_terms(lin, host.scope(), rt1);
+                            walk_terms(lin, host.scope(), rt2);
+                        }
+                        Term::OpN(_, rts) => {
+                            rts.iter().for_each(|rt| walk_terms(lin, host.scope(), rt))
+                        }
+                        Term::StrChunks(chunks) => chunks
+                            .iter()
+                            .filter_map(|chunk| match chunk {
+                                nickel::term::StrChunk::Literal(_) => None,
+                                nickel::term::StrChunk::Expr(rt, _) => Some(rt),
+                            })
+                            .for_each(|rt| walk_terms(lin, host.scope(), rt)),
+                        Term::App(rt1, rt2) => {
+                            walk_terms(lin, host.scope(), rt1);
+                            walk_terms(lin, host.scope(), rt2);
+                        }
+                        Term::Record(rts, _) | Term::RecRecord(rts, _, _) => rts
+                            .values()
+                            .for_each(|rt| walk_terms(lin, host.scope(), rt)),
+                        _ => {}
+                    }
+                }
 
-                for contract in meta.contracts.iter() {
-                    if let nickel::types::AbsType::Flat(RichTerm { term, pos: _ }) =
-                        &contract.types.0
-                    {
-                        if let Term::Var(ident) = &**term {
-                            let parent = self.env.get(ident);
-                            let id = id_gen.get_and_advance();
-                            lin.push(LinearizationItem {
-                                id,
-                                pos: ident.pos.unwrap(),
-                                ty: TypeWrapper::Concrete(AbsType::Var(ident.to_owned())),
-                                scope: self.scope.clone(),
-                                // id = parent: full let binding including the body
-                                // id = parent + 1: actual delcaration scope, i.e. _ = < definition >
-                                kind: TermKind::Usage(UsageState::from(parent)),
-                                meta: None,
-                            });
-                            if let Some(parent) = parent {
-                                lin.add_usage(parent, id);
-                            }
+                // recursively linearize flat types
+                fn walk_types(
+                    lin: &mut Linearization<Building>,
+                    outer_host: &AnalysisHost,
+                    t: &nickel::types::Types,
+                ) {
+                    let inner_host = AnalysisHost {
+                        env: outer_host.env.clone(),
+                        scope: outer_host.scope.clone(),
+                        ..Default::default()
+                    };
```

Comment:
I wonder if it will bite us that we use the same scope and environment here instead of working in their own scope.

Reply:
Ah it will if we have a `let` expression here such as

```
#(let Contract = ... in Contract)
```

contract will leak into the outer environment (only on the linearization but it hurts autocompletion once finally implemented)

fixed in follow-up commit

Reply:
No idea